### PR TITLE
FatFS fixes for F7 series

### PR DESCRIPTION
**Pull request by networkfusion**
## Description

Use our own diskio.c file for FatFS since the ChibiOS project is unwilling to add it.

TODO (as seperate PR's)
Use our own (up to date) FatFS repo
Add read cache invalidation to the chibios_contrib version.

## How Has This Been Tested?
Testing is still to be completed!

## Types of changes

- [ ] Improvement (non-breaking change that improves a feature, code or algorithm)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: targets/CMSIS-OS/ChibiOS/FatFS/fatfs_diskio.c [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+    cacheBufferFlush(buff, MMCSD_BLOCK_SIZE*count);
+    while (count > 0) {
+      if (mmcSequentialRead(&FATFS_HAL_DEVICE, buff))
+        return RES_ERROR;
+      buff += MMCSD_BLOCK_SIZE;
+      count--;
+    }
+    if (mmcStopSequentialRead(&FATFS_HAL_DEVICE))
+        return RES_ERROR;
+    cacheBufferInvalidate(buff, MMCSD_BLOCK_SIZE*count);
```
Comment (josesimoes) [inline targets/CMSIS-OS/ChibiOS/FatFS/fatfs_diskio.c]: I believe this go inside the while above, right after the call to mmcSequentialRead. Depending on the data chunk size that's being read the cache issue may bite us...

File: targets/CMSIS-OS/ChibiOS/FatFS/fatfs_diskio.c [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+#if HAL_USE_MMC_SPI
+  case MMC:
+    if (blkGetDriverState(&FATFS_HAL_DEVICE) != BLK_READY)
+        return RES_NOTRDY;
+    if (mmcIsWriteProtected(&FATFS_HAL_DEVICE))
+        return RES_WRPRT;
+    if (mmcStartSequentialWrite(&FATFS_HAL_DEVICE, sector))
+        return RES_ERROR;
+    // invalidate cache on buffer
+    cacheBufferFlush(buff, MMCSD_BLOCK_SIZE*count);
```
Comment (josesimoes) [inline targets/CMSIS-OS/ChibiOS/FatFS/fatfs_diskio.c]: Same here but inside the while. Similar reason, if we don't flush the cache the data chunk to be written could be still have older data.

File: targets/CMSIS-OS/ChibiOS/FatFS/fatfs_diskio.c (hunk 1 above)
Comment (networkfusion) [reply targets/CMSIS-OS/ChibiOS/FatFS/fatfs_diskio.c]: So if that is the case it needs to lose the *count right...

File: targets/CMSIS-OS/ChibiOS/FatFS/fatfs_diskio.c (hunk 2 above)
Comment (networkfusion) [reply targets/CMSIS-OS/ChibiOS/FatFS/fatfs_diskio.c]: Same here...

**Comment by networkfusion** [comment]
BTW, why is this a breaking change?

File: targets/CMSIS-OS/ChibiOS/FatFS/fatfs_diskio.c (hunk 1 above)
Comment (josesimoes) [reply targets/CMSIS-OS/ChibiOS/FatFS/fatfs_diskio.c]: That is correct. And now that you're mentioning it, on the code above, when it reaches the call to cacheBufferInvalidate(…), the buff pointer is already pointing to the next chunk so no invalidation would actually happen.

The correct place is right after the if and before the call to cacheBufferInvalidate(…)

File: targets/CMSIS-OS/ChibiOS/FatFS/fatfs_diskio.c (hunk 2 above)
Comment (josesimoes) [reply targets/CMSIS-OS/ChibiOS/FatFS/fatfs_diskio.c]: yeap

**Comment by josesimoes** [comment]
> BTW, why is this a breaking change?

It's a breaking change more in the sense that's breaking with ChibiOS code. From a code perspective it looks lile more an enhacement.

File: targets/CMSIS-OS/ChibiOS/FatFS/fatfs_diskio.c (hunk 1 above)
Comment (networkfusion) [reply targets/CMSIS-OS/ChibiOS/FatFS/fatfs_diskio.c]: Not sure I understand... But I will commit what I think is correct and we can have another go if needed...

**Comment by josesimoes** [comment]
Just edited the code to make a couple of adjustments and add proper comments.

**Comment by networkfusion** [comment]
> Just edited the code to make a couple of adjustments and add proper comments.

Fine by me... I am guessing the read commands also need adding to the contrib repo for it to work as expected?

Once this is merged, I will add the changes so that we can seperate from the FatFS zip in chibiOS to using our own git repo...

**Comment by networkfusion** [comment]
Also, I am guessing we need to adjust the FatFS source once separated to we can align buffers?

**Comment by josesimoes** [comment]
> Also, I am guessing we need to adjust the FatFS source once separated to we can align buffers?

That's absolutely mandatory for cache invalidate otherwise will mess up with the adjacent memory.

**Pull request merged**